urllib3/pipeline/base.py

```diff
@@ -0,0 +1,145 @@
+from uuid import uuid4
+
+import six
+
+from urllib3.pipeline.context import Context, ContextWrapper
+
+def has_callable_attr(obj, attr_name):
+    return hasattr(obj, attr_name) and hasattr(getattr(obj, attr_name), '__call__')
+
+
+class HttpPipelineElement(object):
+    
+    def __init__(self):
+        """
+        Each instance of HttpPipelineElement should have a unique identifier
+        that can be used to store state in a context object.
+        """
+        self.id = uuid4().hex
+
+    def save_context(self, context, key, value):
+
+        context.setdefault(self.id, {})
+        context[self.id][key] = value
+
+    def get_context(self, context, key, default=None):
+        return context.get(self.id, {}).get(key, default)
```

Comment:
Not sure I follow - it could lead to confusion if the upstream user is trying to determine if a value they stored is `None`, while the default use will return `None` if nothing's stored, but that could be circumvented by providing an alternate default value.

Would you prefer an implementation that raises a KeyError if a value doesn't exist and a default value isn't explicitly provided? E.g., `self.get_context(context, 'my_key')` would raise, while `self.get_context(context, 'my_key', None)` would return `None`.